### PR TITLE
option to not delay TCPServer socket creation

### PR DESCRIPTION
**Pull request by tanmaykm**
This change adds an option to TCPServer socket constructor to not delay creation of the socket. This will allow TCP servers to set socket options like `SO_REUSEPORT` before the socket is bound and listened on.

With early initialization:

```
tan@tanlt ~/temp/soreuseport $ julia ./soreuseport.jl false &
[1] 23061
tan@tanlt ~/temp/soreuseport $ creating socket bound to 8010...
has reuseport
tcp_reuseport returned 0
listening...
accepting...

tan@tanlt ~/temp/soreuseport $ julia ./soreuseport.jl false
creating socket bound to 8010...
has reuseport
tcp_reuseport returned 0
listening...
accepting...
```

But that is not possible with delayed initialization:

```
tan@tanlt ~/temp/soreuseport $ julia ./soreuseport.jl true &
[1] 22977
tan@tanlt ~/temp/soreuseport $ creating socket bound to 8010...
has reuseport
tcp_reuseport returned -1
listening...
accepting...

tan@tanlt ~/temp/soreuseport $ julia ./soreuseport.jl true
creating socket bound to 8010...
has reuseport
tcp_reuseport returned -1
listening...
ERROR: LoadError: listen: address already in use (EADDRINUSE)
Stacktrace:
 [1] uv_error at ./libuv.jl:68 [inlined]
 [2] #listen#369 at ./stream.jl:933 [inlined]
 [3] listen at ./stream.jl:933 [inlined]
 [4] run_server(::Bool) at /home/tan/temp/soreuseport/soreuseport.jl:13
 [5] include_relative(::Module, ::String) at ./loading.jl:533
 [6] include(::Module, ::String) at ./sysimg.jl:14
 [7] process_options(::Base.JLOptions) at ./client.jl:325
 [8] _start() at ./client.jl:391
in expression starting at /home/tan/temp/soreuseport/soreuseport.jl:22
```

My `soreuseport.jl` is:

```
function run_server(delay)
    println("creating socket bound to 8010...")
    s = Base.TCPServer(; delay=delay)
    if ccall(:jl_has_so_reuseport, Int32, ()) == 1
        println("has reuseport")
        rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Void},), s.handle)
        println("tcp_reuseport returned ", rc)
    else
        println("no reuseport")
    end
    bind(s, Base.InetAddr(ip"0.0.0.0", 8010))
    println("listening...")
    listen(s)
    println("accepting...")
    s1 = accept(s)
    println("accepted, closing...")
    close(s1)
    close(s)
    println("done")
end

run_server(parse(Bool, ARGS[1]))
```

### CODE REVIEW

**Review by StefanKarpinski — COMMENTED** [review]
LGTM. Needs docs and ideally a test even if only to test that the keyword is accepted, if not that it does something in particular.

File: test/socket.jl [hunk 1]
```diff
@@ -407,3 +407,10 @@ end
         @test test_connect(addr)
     end
 end
+
+@testset "TCPServer constructor" begin
+    s = Base.TCPServer(; delay=false)
+    if ccall(:jl_has_so_reuseport, Int32, ()) == 1
+        @test 0 == ccall(:jl_tcp_reuseport, Int32, (Ptr{Void},), s.handle)
+    end
+end
```
Comment (StefanKarpinski) [inline test/socket.jl]: Maybe test the positive case as well? (Unless that's already done somewhere else)

File: base/socket.jl [hunk 2]
```diff
@@ -312,6 +312,10 @@ mutable struct TCPServer <: LibuvServer
         return tcp
     end
 end
+
+# Keyword arg "delay": if true, libuv delays creation of socket fd till bind.
+# It can be set to false if there is a need to set socket options before
+# further calls to `bind` and `listen`, e.g. `SO_REUSEPORT`.
```
Comment (StefanKarpinski) [inline base/socket.jl]: I meant as part of the doc string for the `TCPServer` constructor, although a comment is better than nothing.

Comment (tanmaykm) [reply base/socket.jl]: TCPServer is not exported, so I think this should be okay for now.

File: test/socket.jl (hunk 1 above)
Comment (tanmaykm) [reply test/socket.jl]: Yes, that is tested elsewhere as that's the default.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by tanmaykm** [comment]
Can this be merged now?
It will be good to have this backported to 0.6 as well.

**Comment by aviks** [comment]
Bump, anything preventing this from being merged?

**Comment by aviks** [comment]
@ararslan Any chance getting this into 0.6.2 please?

**Comment by ararslan** [comment]
Yes, will do.

**Comment by tkelman** [comment]
A new keyword argument is a feature, not a bugfix.

**Comment by tanmaykm** [comment]
`TCPServer` constructor is not exported, and there is no change in the default behavior.

**Comment by tkelman** [comment]
So? Shouldn't encourage writing code that works on 0.6.2 that wouldn't work on 0.6.1 unless there's a legitimate bug involved.

**Comment by StefanKarpinski** [comment]
Strictly, @tkelman is right, of course, but @tanmaykm is also right that this is a new option to an unexported type, which is about as mild as a feature gets. What's at stake here? What is this feature needed for? Is this effectively a bug fix?

**Comment by aviks** [comment]
So this feature is needed to run a Julia web server with real world load. Currently, a Julia web server can effectively only serve one connection at a time (even though the webserver is written with `@async`). This allows us to serve multiple connections on one machine. So in some ways this is crucial for realistic deploys. 

The actual underlying feature (`SO_REUSEPORT`) has been in Julia (via libuv) for a long time. However, it cannot be used, due to an incorrect order of initialisation (to simplify things a bit). This PR changes the order of initialisation based on the keyword argument, thus allowing the feature to be actually used. So yes, this could be seen as a bugfix, since the feature has been added, but was unable to be used. 

I think we can safely say that a new keyword argument is pretty low risk, there are bugfixes with much higher risk (as we've seen recently). Hence I would argue for its inclusion.

**Comment by StefanKarpinski** [comment]
Could we just change the incorrect initialization order and consider that a bugfix?

**Comment by tkelman** [comment]
Or carry a local copy of the modified version of the constructor, which should work correctly even on 0.6.0 or 0.6.1.

**Comment by tanmaykm** [comment]
To answer Stefan's question, I consider the keyword argument as the proper fix.

The bug was that Julia allowed only one way of constructing server socket (delayed), where no options could be set until it is listened on. While delayed creation is useful to avoid locking resources until use, we also need early creation to be able to set certain options.

Without the keyword we will have to make all sockets initialize early, which I think is worse.

**Comment by tanmaykm** [comment]
`TCPServer()` is an internal API. Exported APIs `listen` and `listenany` are the only ones that use this. The next step should be to have these exported APIs changed. I would agree with not backporting those to `v0.6`.

I would however argue for backporting this. It would allow packages that are willing to depend on this internal API in `> v0.6.2` onwards to get this feature without calling a bunch of other internal APIs.

There is no impact on packages that choose to be compatible across `v0.6`. They can either not use reuseport at all and be oblivious of this change, or implement all of the constructor on their own.

**Comment by tkelman** [comment]
There shouldn't be such a thing as an API that only exists in 0.6.2, internal or otherwise. Feature freeze for release-0.6 was in February.

**Pull request merged**